packages/databases-collections/src/modules/create-collection/index.js

```diff
@@ -117,35 +118,84 @@ export const createCollection = (data, kind = 'Collection') => {
-          is_capped: !!data.options.capped,
-          has_collation: !!data.options.collation,
-          is_timeseries: !!data.options.timeseries,
-          is_clustered: !!data.options.clusteredIndex,
-          expires: !!data.options.expireAfterSeconds
-        };
 
-        track(`${kind} Created`, trackEvent);
+      const options = await handleFLE2Options(ds, data.options);
 
-        prepareMetrics(collection).then((metrics) => {
-          global.hadronApp.appRegistry.emit('compass:collection:created', metrics);
-        });
-        global.hadronApp.appRegistry.emit('collection-created', {
-          ns: namespace,
-          database: dbName,
-          collection: collName,
+      const collection = await new Promise((resolve, reject) => {
+        ds.createCollection(namespace, options, (err, coll) => {
+          if (err) reject(err); else resolve(coll);
         });
-        dispatch(reset());
       });
+
+      const trackEvent = {
+        is_capped: !!data.options.capped,
+        has_collation: !!data.options.collation,
+        is_timeseries: !!data.options.timeseries,
+        is_clustered: !!data.options.clusteredIndex,
+        expires: !!data.options.expireAfterSeconds
+      };
+
+      track(`${kind} Created`, trackEvent);
+
+      prepareMetrics(collection).then((metrics) => {
+        global.hadronApp.appRegistry.emit('compass:collection:created', metrics);
+      });
+      global.hadronApp.appRegistry.emit('collection-created', {
+        ns: namespace,
+        database: dbName,
+        collection: collName,
+      });
+      dispatch(reset());
     } catch (e) {
       return stopWithError(dispatch, e);
     }
   };
 };
 
+export async function handleFLE2Options(ds, options) {
+  if (!options) {
+    return options;
+  }
+
+  if (options.encryptedFields) {
+    try {
+      options.encryptedFields = queryParser(options.encryptedFields);
+    } catch (err) {
+      throw new Error(`Could not parse encryptedFields config: ${err.message}`);
+    }
+
+    if (Object.keys(options.encryptedFields).length === 0) {
+      delete options.encryptedFields;
+    } else if (options.kmsProvider) {
+      // If keys are missing from the encryptedFields config,
+      // generate them as part of the collection creation operation.
+      let keyEncryptionKey;
+      try {
+        keyEncryptionKey = queryParser(options.keyEncryptionKey || '{}');
+      } catch (err) {
+        throw new Error(`Could not parse keyEncryptionKey: ${err.message}`);
+      }
+
+      const fields = options.encryptedFields.fields;
+      if (Array.isArray(fields)) {
+        const keyCreationPromises = [];
+        for (const field of fields) {
+          if (field.keyId) continue;
+          keyCreationPromises.push((async() => {
+            field.keyId = await ds.createDataKey(options.kmsProvider, {
+              masterKey: keyEncryptionKey
+            });
+          })());
+        }
+        await Promise.all(keyCreationPromises);
```

Comment:
Automatically creating data keys when they were missing from the `encryptedFields` config was cut from the the drivers FLE2 support project, so we have to do it manually here. Since it isn’t that much work to add it in Compass itself, and it is part of enabling users to get started with FLE2 without having to leave the Compass UI, I think we should still keep doing it. (The alternative would be requiring users to manually create keys in the shell, then copy the key UUIDs to the `encryptedFields` editor.)